lib/src/main/java/com/payone/commerce/platform/lib/endpoints/BaseApiClient.java

```diff
@@ -0,0 +1,110 @@
+package com.payone.commerce.platform.lib.endpoints;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.payone.commerce.platform.lib.CommunicatorConfiguration;
+import com.payone.commerce.platform.lib.RequestHeaderGenerator;
+import com.payone.commerce.platform.lib.errors.ApiErrorResponseException;
+import com.payone.commerce.platform.lib.errors.ApiResponseRetrievalException;
+import com.payone.commerce.platform.lib.models.ErrorResponse;
+
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class BaseApiClient {
+
+    private final OkHttpClient client = new OkHttpClient();
+    private final String JSON_PARSE_ERROR = "Excepted valid JSON response, but failed to parse";
+    protected final JsonMapper JSON_MAPPER = new JsonMapper();
+    protected final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
+
+    private final RequestHeaderGenerator requestHeaderGenerator;
+    private final CommunicatorConfiguration config;
+
+    public BaseApiClient(CommunicatorConfiguration config) throws InvalidKeyException {
+        this.config = config;
+        this.requestHeaderGenerator = new RequestHeaderGenerator(config);
+        JSON_MAPPER.registerModule(new JavaTimeModule());
+
+    }
+
+    public BaseApiClient() {
+        this.config = null;
+        this.requestHeaderGenerator = null;
+    }
+
+    protected RequestHeaderGenerator getRequestHeaderGenerator() {
+        return requestHeaderGenerator;
+    }
+
+    protected OkHttpClient getClient() {
+        return client;
+    }
+
+    protected CommunicatorConfiguration getConfig() {
+        return config;
+    }
+
+    protected JsonMapper getJsonMapper() {
+        return JSON_MAPPER;
+    }
+
+    protected void makeApiCall(Request request)
+            throws IOException, ApiErrorResponseException, ApiResponseRetrievalException {
+        Response response = getResponse(request);
+        handleError(response);
+    }
+
+    protected <T> T makeApiCall(Request request, TypeReference<T> valueTypeRef)
+            throws IOException, ApiErrorResponseException, ApiResponseRetrievalException {
+        Response response = getResponse(request);
```

Comment:
Wir könnten den `requestHeaderGenerator` auch hier aufrufen. Da müssen wir auch nicht daran denken, dass in jedem Endpunkt nochmal zu machen. Soweit ich das sehe, machen wir das eh immer direkt vor dem `makeApiCall`-Aufruf.